contrib/parseq-restli-client/src/main/java/com/linkedin/restli/client/ParSeqRestClient.java

```diff
@@ -220,44 +220,51 @@ static String generateTaskName(final Request<?> request) {
-    }
-    requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT, d2Timeout);
-    return d2Timeout;
+    return (requestTimeout instanceof Number) && (((Number)requestTimeout).intValue() > 0);
   }
 
+  /**
+   * We will distinguish two cases in applying timeout to a ParSeq rest.li request task.
+   * Case 1: There is no per request timeout specified in request context of rest.li request, timeout is configured
+   * through ParSeqRestClient configuration. For this case, we will update request context as:
+   *    REQUEST_TIMEOUT = configured timeout value
+   *    REQUEST_TIMEOUT_IGNORE_IF_HIGHER_THAN_DEFAULT = true
+   * since in this case, ParSeqRestClient just wants to timeout this request from client side within configured timeout
+   * without disturbing any lower layer load balancing behaviors.
+   *
+   * Case 2: There is per request timeout specified in rest.li request, and there may or may not have timeout specified
+   * through ParSeqRestClient configuration. For this case, per request timeout specified in rest.li request always
+   * takes precedence, ParSeq will interpret that users would like to use this to impact lower layer LB behavior, and
+   * thus will pass down request context unchanged down.
+   */
   private <T> Task<Response<T>> createTaskWithTimeout(final String name, final Request<T> request,
       final RequestContext requestContext, RequestConfig config) {
     ConfigValue<Long> timeout = config.getTimeoutMs();
-    int reconciledTimeout = 0;
-    if (timeout.getValue() != null && timeout.getValue() > 0 && _d2RequestTimeoutEnabled) {
-      // reconcile timeout if request has already set timeout in its request context
-      reconciledTimeout = reconcileRequestTimeout(requestContext, timeout.getValue());
+    int configuredTimeout = timeout.getValue().intValue();
+    boolean hasPerRequestTimeout = hasRequestContextTimeout(requestContext);
+    if (timeout.getValue() != null && timeout.getValue() > 0 && !hasPerRequestTimeout && _d2RequestTimeoutEnabled) {
+      // timeout is configured and no per-request timeout specified in the request context
+      requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT, configuredTimeout);
+      requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT_IGNORE_IF_HIGHER_THAN_DEFAULT, true);
     }
     Task<Response<T>> requestTask;
     if (RequestGroup.isBatchable(request, config)) {
       requestTask = createBatchableTask(name, request, requestContext, config);
     } else {
       requestTask = Task.async(name, () -> sendRequest(request, requestContext));
     }
-    if (timeout.getValue() == null || timeout.getValue() <= 0) {
-      // no timeout configured
+    if (timeout.getValue() == null || timeout.getValue() <= 0 || hasPerRequestTimeout) {
+      // no timeout configured or per-request timeout already specified in request context
       return requestTask;
     } else {
       if (!_d2RequestTimeoutEnabled) {
```

Comment:
Seems that if `d2RequestTimeout` is not enabled, it will just take out the timeout value and call `withTimeout`. This function looks really messy because of the need to put the right timeout in the requestContext. Can we toggle on `d2RequestTimeoutEnabled` and have two separate functions to improve cleanness?

Reply:
Thanks for suggestion, I will split them into two functions to make it cleaner.